CadRevealFbxProvider/BatchUtils/ScaffoldOptimizer.cs

```diff
@@ -0,0 +1,95 @@
+namespace CadRevealFbxProvider.BatchUtils;
+using CadRevealComposer;
+using CadRevealComposer.Primitives;
+using CadRevealComposer.Tessellation;
+using System.Linq;
+using ScaffoldPartOptimizers;
+
+public static class ScaffoldOptimizer
+{
+    public static void AddPartOptimizer(ScaffoldPartOptimizer optimizer)
+    {
+        PartOptimizers.Add(optimizer);
+    }
+
+    public static void OptimizeNode(CadRevealNode node)
+    {
+        var name = node.Name;
+
+        var newGeometries = new List<APrimitive>();
+        foreach (APrimitive primitive in node.Geometries)
+        {
+            newGeometries.AddRange(OptimizePrimitive(primitive, name));
+        }
+
+        node.Geometries = newGeometries.ToArray();
+    }
+
+    private static APrimitive[] OptimizePrimitive(APrimitive primitive, string name)
+    {
+        // Handle only primitives that have their own Mesh objects, then pull out the mesh and optimize. These are the ones that need to be optimized.
+        var primitiveList = new List<APrimitive>();
+        switch (primitive)
+        {
+            case InstancedMesh instancedMesh:
+                {
+                    Mesh[] meshes = OptimizeMesh(instancedMesh.TemplateMesh, name);
+                    primitiveList.AddRange(meshes.Select(mesh => new InstancedMesh(instancedMesh.InstanceId, mesh,
+                        instancedMesh.InstanceMatrix, instancedMesh.TreeIndex, instancedMesh.Color, instancedMesh.AxisAlignedBoundingBox)));
+                    break;
+                }
+            case TriangleMesh triangleMesh:
+                {
+                    Mesh[] meshes = OptimizeMesh(triangleMesh.Mesh, name);
+                    primitiveList.AddRange(meshes.Select(mesh => new TriangleMesh(mesh, triangleMesh.TreeIndex, triangleMesh.Color,
+                        triangleMesh.AxisAlignedBoundingBox)));
+                    break;
+                }
+        }
+
+        return primitiveList.ToArray();
```

Comment:
This will return an empty list if a non-mesh primitive is sent in. Consider throwing an exception to catch this if it happens instead of failing silently.

Reply:
I believe this should not throw an exception, but rather return a nullable list. This is because all scaffold nodes will call OptimizeNode, leading to a high probability of encountering non-mesh primitives. Thus, the current code would likely fail. However, by returning a nullable it is possible to simply pass the primitive through if it is not a mesh-primitive. I think I should also add some non-mesh primitive to the tests, since the tests did not pick up on the problem in the first place.